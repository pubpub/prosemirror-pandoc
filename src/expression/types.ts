export type Identifier = {
    type: "identifier";
    identifier: string;
};

export type OneOrMore = {
    type: "oneOrMore";
    child: Expr;
};

export type ZeroOrMore = {
    type: "zeroOrMore";
    child: Expr;
};

export type Range = {
    type: "range";
    lowerBound: number;
    upperBound: number | null;
    child: Expr;
};

export type Sequence = {
    type: "sequence";
    children: Expr[];
};
export type Choice = {
    type: "choice";
    children: Expr[];
};

export type IdentifierMatch<Item> = (id: string) => (item: Item) => boolean;

export type Expr =
    | Identifier
    | OneOrMore
    | ZeroOrMore
    | Sequence
    | Choice
    | Range;
