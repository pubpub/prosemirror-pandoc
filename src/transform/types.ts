import {
    PandocNode,
    ProsemirrorElement,
    ProsemirrorMark,
    ProsemirrorNode,
    ProsemirrorSchema,
} from "../types";
import { ProsemirrorFluent, PandocFluent } from "./fluent";
import { Expr } from "./nodeExpression";

export type Transformer<From, To> = (
    node: From,
    context: From extends PandocNode
        ? TransformContext<PandocNode, ProsemirrorNode>
        : TransformContext<ProsemirrorNode, PandocNode>
) => To | To[];

export type BidiTransformer<
    PdNode extends PandocNode = PandocNode,
    PmElem extends ProsemirrorElement = ProsemirrorElement
> = {
    fromPandoc: Transformer<PdNode, PmElem>;
    fromProsemirror: Transformer<PmElem, PdNode>;
};

export type WrappedBidiTransformer<
    PdNode extends PandocNode = PandocNode,
    PmElem extends ProsemirrorElement = ProsemirrorElement
> =
    | BidiTransformer<PdNode, PmElem>
    | ((
          pdNodeName: string,
          pmNodeName: string
      ) => BidiTransformer<PdNode, PmElem>);

export type TransformConfig = {
    resource: (input: string, context?: any) => string;
    useSmartQuotes: boolean;
    prosemirrorDocWidth: number;
};

export type TransformParentContext = {
    textAlign?: "left" | "center" | "right";
};

export type TransformContext<From, To> = TransformConfig &
    TransformParentContext & {
        transform: (
            from: From | From[],
            options?: {
                context?: Partial<TransformParentContext>;
                marks?: ProsemirrorMark[];
            }
        ) => To extends PandocNode ? PandocFluent : ProsemirrorFluent;
        rules: Rule<From, To>[];
        count: (nodeType: string) => number;
        marksMap: Map<To, ProsemirrorMark[]>;
        prosemirrorSchema: ProsemirrorSchema;
    };

export type Rule<From, To> = {
    transform: Transformer<From | From[], To>;
    expression: Expr;
    acceptsMultiple: boolean;
};

export type RuleSet = {
    fromPandoc: Rule<PandocNode, ProsemirrorNode>[];
    fromProsemirror: Rule<ProsemirrorElement, PandocNode>[];
    prosemirrorSchema: ProsemirrorSchema;
};
