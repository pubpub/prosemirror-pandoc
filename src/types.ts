/**
 * Definitions for the Pandoc AST
 * See https://hackage.haskell.org/package/pandoc-types-1.22/docs/Text-Pandoc-Definition.html
 */

export { Schema as ProsemirrorSchema } from "prosemirror-model";

export type ProsemirrorAttr =
    | undefined
    | null
    | number
    | string
    | ProsemirrorAttr[];

export type ProsemirrorNode<Type = string> = {
    __isMark?: false;
    type: Type;
    content?: ProsemirrorNode[];
    text?: string;
    attrs?: Record<string, ProsemirrorAttr>;
    marks?: ProsemirrorMark[];
};

export type ProsemirrorMark<Type = string> = {
    __isMark?: true;
    type: Type;
    attrs?: Record<string, ProsemirrorAttr>;
};

export type ProsemirrorElement = ProsemirrorNode | ProsemirrorMark;

export type PandocJson = {
    "pandoc-api-version": number[];
    meta: {
        [key: string]: any;
    };
    blocks: { [key: string]: any }[];
};
export type Doc = {
    type: "Doc";
    blocks: Block[];
    meta: {
        [key: string]: MetaValue;
    };
};

export type Alignment =
    | "AlignLeft"
    | "AlignRight"
    | "AlignCenter"
    | "AlignDefault";

export type QuoteType = "SingleQuote" | "DoubleQuote";
export type MathType = "DisplayMath" | "InlineMath";

export type ListNumberStyle =
    | "DefaultStyle"
    | "Example"
    | "Decimal"
    | "LowerRoman"
    | "UpperRoman"
    | "LowerAlpha"
    | "UpperAlpha";

export type ListNumberDelim =
    | "DefaultDelim"
    | "Period"
    | "OneParen"
    | "TwoParens";

export type ListAttributes = {
    startNumber: number;
    listNumberStyle: ListNumberStyle;
    listNumberDelim: ListNumberDelim;
};

export type Format = string;

export type Attr = {
    identifier: string;
    classes: string[];
    properties: { [key: string]: string };
};

export type Target = {
    url: string;
    title: string;
};

export type CitationMode = "AuthorInText" | "SuppressAuthor" | "NormalCitation";

export type Citation = {
    citationId: string;
    citationPrefix: Inline[];
    citationSuffix: Inline[];
    citationMode: CitationMode;
    citationNoteNum: number;
    citationHash: number;
};

/* ~~~ Block-level definitions ~~~ */

/* Plain text, not a paragraph */
export type Plain = {
    type: "Plain";
    content: Inline[];
};

/* Paragraph */
export type Para = {
    type: "Para";
    content: Inline[];
};

/* Multiple non-breaking lines */
export type LineBlock = {
    type: "LineBlock";
    content: Inline[][];
};

/* Code block (literal) with attributes */
export type CodeBlock = {
    type: "CodeBlock";
    attr: Attr;
    content: string;
};

/* Raw block */
export type RawBlock = {
    type: "RawBlock";
    format: Format;
    content: string;
};

/* Block quote (list of blocks) */
export type BlockQuote = {
    type: "BlockQuote";
    content: Block[];
};

/* Ordered list (attributes and a list of items, each a list of blocks) */
export type OrderedList = {
    type: "OrderedList";
    listAttributes: ListAttributes;
    content: Block[][];
};

/* Bullet list (list of items, each a list of blocks) */
export type BulletList = {
    type: "BulletList";
    content: Block[][];
};

/* Definition list
 Each list item is a pair consisting of a term (a list of inlines)
 and one or more definitions (each a list of blocks) */
export type DefinitionList = {
    type: "DefinitionList";
    entries: {
        term: Inline[];
        definitions: Block[][];
    }[];
};

/* Header - level (integer) and text (inlines) */
export type Header = {
    type: "Header";
    level: number;
    attr: Attr;
    content: Inline[];
};

/* Horizontal rule */
export type HorizontalRule = {
    type: "HorizontalRule";
};

/* Table stuff */
export type Caption = {
    type: "Caption";
    shortCaption?: Inline[];
    content: Block[];
};

export type ColSpec = {
    type: "ColSpec";
    alignment: Alignment;
} & ({ width: number } | { defaultWidth: true });

export type Cell = {
    type: "Cell";
    attr: Attr;
    alignment: Alignment;
    rowSpan: number;
    colSpan: number;
    content: Block[];
};

export type Row = {
    type: "Row";
    attr: Attr;
    cells: Cell[];
};

export type TableHead = {
    type: "TableHead";
    attr: Attr;
    rows: Row[];
};

export type TableFoot = {
    type: "TableFoot";
    attr: Attr;
    rows: Row[];
};

export type TableBody = {
    type: "TableBody";
    attr: Attr;
    rowHeadColumns: number;
    headRows: Row[];
    bodyRows: Row[];
};

export type Table = {
    type: "Table";
    attr: Attr;
    caption: Caption;
    colSpecs: ColSpec[];
    head: TableHead;
    bodies: TableBody[];
    foot: TableFoot;
};

export type TableRow = {
    attr: Attr;
    rowHeadColumns: number;
};

/* Generic block container with attributes */
export type Div = {
    type: "Div";
    attr: Attr;
    content: Block[];
};

/* Nothing */
export type Null = {
    type: "Null";
};

export type Block =
    | Plain
    | Para
    | LineBlock
    | CodeBlock
    | RawBlock
    | BlockQuote
    | OrderedList
    | BulletList
    | DefinitionList
    | Header
    | HorizontalRule
    | Table
    | Div
    | Null;

/* ~~~ Inline-level definitions ~~~ */

/* Text (string) */
export type Str = {
    type: "Str";
    content: string;
};

/* Emphasized text (list of inlines) */
export type Emph = {
    type: "Emph";
    content: Inline[];
};

/* Underlined text (list of inlines) */
export type Underline = {
    type: "Underline";
    content: Inline[];
};

/* Strongly emphasized text (list of inlines) */
export type Strong = {
    type: "Strong";
    content: Inline[];
};

/* Strikeout text (list of inlines) */
export type Strikeout = {
    type: "Strikeout";
    content: Inline[];
};

/* Superscripted text (list of inlines) */
export type Superscript = {
    type: "Superscript";
    content: Inline[];
};

/* Subscripted text (list of inlines) */
export type Subscript = {
    type: "Subscript";
    content: Inline[];
};

/* Small caps text (list of inlines) */
export type SmallCaps = {
    type: "SmallCaps";
    content: Inline[];
};

/* Quoted text (list of inlines) */
export type Quoted = {
    type: "Quoted";
    quoteType: QuoteType;
    content: Inline[];
};

/* Citation (list of inlines) */
export type Cite = {
    type: "Cite";
    citations: Citation[];
    content: Inline[];
};

/* Inline code (literal) */
export type Code = {
    type: "Code";
    attr: Attr;
    content: string;
};

/* Inter-word space */
export type Space = {
    type: "Space";
};

/* Soft line break */
export type SoftBreak = {
    type: "SoftBreak";
};

/* Hard line break */
export type LineBreak = {
    type: "LineBreak";
};

/* TeX math (literal) */
export type Math = {
    type: "Math";
    mathType: MathType;
    content: string;
};

/* Raw inline */
export type RawInline = {
    type: "RawInline";
    format: Format;
    content: string;
};

/* Hyperlink: alt text (list of inlines), target */
export type Link = {
    type: "Link";
    attr: Attr;
    content: Inline[];
    target: Target;
};

/* Image: alt text (list of inlines), target */
export type Image = {
    type: "Image";
    attr: Attr;
    content: Inline[];
    target: Target;
};

/* Footnote or endnote */
export type Note = {
    type: "Note";
    content: Block[];
};

/* Generic inline container with attributes */
export type Span = {
    type: "Span";
    attr: Attr;
    content: Inline[];
};

/* Meta types */

export type MetaMap = {
    type: "MetaMap";
    values: { [key: string]: MetaValue };
};

export type MetaList = {
    type: "MetaList";
    content: MetaValue[];
};

export type MetaBool = {
    type: "MetaBool";
    content: boolean;
};

export type MetaString = {
    type: "MetaString";
    content: string;
};
export type MetaInlines = {
    type: "MetaInlines";
    content: Inline[];
};

export type MetaBlocks = {
    type: "MetaBlocks";
    content: Block[];
};

export type MetaValue =
    | MetaMap
    | MetaList
    | MetaBool
    | MetaString
    | MetaInlines
    | MetaBlocks;

export type SimpleInline =
    | Emph
    | Underline
    | Strong
    | Strikeout
    | Superscript
    | Subscript
    | SmallCaps;

export type Inline =
    | Str
    | SimpleInline
    | Quoted
    | Cite
    | Code
    | Space
    | SoftBreak
    | LineBreak
    | Math
    | RawInline
    | Link
    | Image
    | Note
    | Span;

export type PandocNode = Doc | Block | Inline;

export const PANDOC_NODE_TYPES = [
    "BlockQuote",
    "BulletList",
    "Cite",
    "Code",
    "CodeBlock",
    "DefinitionList",
    "Doc",
    "Div",
    "Emph",
    "Header",
    "HorizontalRule",
    "Image",
    "LineBlock",
    "LineBreak",
    "Link",
    "Math",
    "Note",
    "Null",
    "OrderedList",
    "Para",
    "Plain",
    "Quoted",
    "RawBlock",
    "RawInline",
    "SmallCaps",
    "SoftBreak",
    "Space",
    "Span",
    "Str",
    "Strikeout",
    "Strong",
    "Subscript",
    "Superscript",
    "Table",
    "Underline",
];
