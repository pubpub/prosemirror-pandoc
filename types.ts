/**
 * Definitions for the Pandoc AST
 * See https://hackage.haskell.org/package/pandoc-types-1.17.6/docs/Text-Pandoc-Definition.html
 */

export type BlockNodeType =
    | "Plain"
    | "Para"
    | "LineBlock"
    | "CodeBlock"
    | "RawBlock"
    | "BlockQuote"
    | "OrderedList"
    | "BulletList"
    | "DefinitionList"
    | "Header"
    | "HorizontalRule"
    | "Table"
    | "Div"
    | "Null";

export type InlineNodeType =
    | "Str"
    | "Emph"
    | "Strong"
    | "Strikeout"
    | "Superscript"
    | "Subscript"
    | "SmallCaps"
    | "Quoted"
    | "Cite"
    | "Code"
    | "Space"
    | "SoftBreak"
    | "LineBreak"
    | "Math"
    | "RawInline"
    | "Link"
    | "Image"
    | "Note"
    | "Span";

export interface Doc {
    blocks: Block[];
    meta: {};
}

export type NodeType = BlockNodeType | InlineNodeType;

export interface Node {
    type: NodeType;
    content?: any[] | string;
}

export interface Block extends Node {
    type: BlockNodeType;
}

export interface Inline extends Node {
    type: InlineNodeType;
}

export type Alignment =
    | "AlignLeft"
    | "AlignRight"
    | "AlignCenter"
    | "AlignDefault";

export type QuoteType = "SingleQuote" | "DoubleQuote";
export type MathType = "DisplayMath" | "InlineMath";

export type ListNumberStyle =
    | "DefaultStyle"
    | "Example"
    | "Decimal"
    | "LowerRoman"
    | "UpperRoman"
    | "LowerAlpha"
    | "UpperAlpha";

export type ListNumberDelim =
    | "DefaultDelim"
    | "Period"
    | "OneParen"
    | "TwoParens";

export interface ListAttributes {
    startNumber: number;
    listNumberStyle: ListNumberStyle;
    listNumberDelim: ListNumberDelim;
}

export type Format = string;

export interface Attr {
    identifier: string;
    classes: string[];
    properties: { [key: string]: string };
}

export interface Target {
    url: string;
    title: string;
}

export type CitationMode = "AuthorInText" | "SuppressAuthor" | "NormalCitation";

export interface Citation {
    citationId: string;
    citationPrefix: Inline[];
    citationSuffix: Inline[];
    citationMode: CitationMode;
    citationNoteNum: number;
    citationHash: number;
}

/* ~~~ Block-level definitions ~~~ */

/* Plain text, not a paragraph */
export interface Plain extends Block {
    type: "Plain";
    content: Inline[];
}

/* Paragraph */
export interface Para extends Block {
    type: "Para";
    content: Inline[];
}

/* Multiple non-breaking lines */
export interface LineBlock extends Block {
    type: "LineBlock";
    content: Inline[][];
}

/* Code block (literal) with attributes */
export interface CodeBlock extends Block {
    type: "CodeBlock";
    attr: Attr;
    content: string;
}

/* Raw block */
export interface RawBlock extends Block {
    type: "RawBlock";
    format: Format;
    content: string;
}

/* Block quote (list of blocks) */
export interface BlockQuote extends Block {
    type: "BlockQuote";
    content: Block[];
}

/* Ordered list (attributes and a list of items, each a list of blocks) */
export interface OrderedList extends Block {
    type: "OrderedList";
    listAttributes: ListAttributes;
    content: Block[][];
}

/* Bullet list (list of items, each a list of blocks) */
export interface BulletList extends Block {
    type: "BulletList";
    content: Block[][];
}

/* Definition list
 Each list item is a pair consisting of a term (a list of inlines)
 and one or more definitions (each a list of blocks) */
export interface DefinitionList extends Block {
    type: "DefinitionList";
    terms: Inline[];
    definitions: Block[][];
}

/* Header - level (integer) and text (inlines) */
export interface Header extends Block {
    type: "Header";
    level: number;
    attr: Attr;
    content: Inline[];
}

/* Horizontal rule */
export interface HorizontalRule extends Block {
    type: "HorizontalRule";
}

/* Table, with caption, column alignments (required),
 relative column widths (0 = default), column headers (each a list of blocks),
 and rows (each a list of lists of blocks) */
export interface Table extends Block {
    type: "Table";
    caption: Inline[];
    alignments: Alignment[];
    columnWidths: number[];
    headers: Block[][];
    cells: Block[][][];
}

/* Generic block container with attributes */
export interface Div extends Block {
    type: "Div";
    attr: Attr;
    content: Block[];
}

/* Nothing */
export interface Null extends Block {
    type: "Null";
}

/* ~~~ Inline-level definitions ~~~ */

/* Text (string) */
export interface Str extends Inline {
    type: "Str";
    content: string;
}

/* Emphasized text (list of inlines) */
export interface Emph extends Inline {
    type: "Emph";
    content: Inline[];
}

/* Strongly emphasized text (list of inlines) */
export interface Strong extends Inline {
    type: "Strong";
    content: Inline[];
}

/* Strikeout text (list of inlines) */
export interface Strikeout extends Inline {
    type: "Strikeout";
    content: Inline[];
}

/* Superscripted text (list of inlines) */
export interface Superscript extends Inline {
    type: "Superscript";
    content: Inline[];
}

/* Subscripted text (list of inlines) */
export interface Subscript extends Inline {
    type: "Subscript";
    content: Inline[];
}

/* Small caps text (list of inlines) */
export interface SmallCaps extends Inline {
    type: "SmallCaps";
    content: Inline[];
}

/* Quoted text (list of inlines) */
export interface Quoted extends Inline {
    type: "Quoted";
    quoteType: QuoteType;
    content: Inline[];
}

/* Citation (list of inlines) */
export interface Cite extends Inline {
    type: "Cite";
    citations: Citation[];
    content: Inline[];
}

/* Inline code (literal) */
export interface Code extends Inline {
    type: "Code";
    attr: Attr;
    content: string;
}

/* Inter-word space */
export interface Space extends Inline {
    type: "Space";
}

/* Soft line break */
export interface SoftBreak extends Inline {
    type: "SoftBreak";
}

/* Hard line break */
export interface LineBreak extends Inline {
    type: "LineBreak";
}

/* TeX math (literal) */
export interface Math extends Inline {
    type: "Math";
    mathType: MathType;
    content: string;
}

/* Raw inline */
export interface RawInline extends Inline {
    type: "RawInline";
    format: Format;
    content: string;
}

/* Hyperlink: alt text (list of inlines), target */
export interface Link extends Inline {
    type: "Link";
    attr: Attr;
    content: Inline[];
    target: Target;
}

/* Image: alt text (list of inlines), target */
export interface Image extends Inline {
    type: "Image";
    attr: Attr;
    content: Inline[];
    target: Target;
}

/* Footnote or endnote */
export interface Note extends Inline {
    type: "Note";
    content: Block[];
}

/* Generic inline container with attributes */
export interface Span extends Inline {
    type: "Span";
    attr: Attr;
    content: Inline[];
}
