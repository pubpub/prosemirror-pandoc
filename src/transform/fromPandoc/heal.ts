import { ProsemirrorNode, ProsemirrorSchema } from "types";
import { parseExpr, Expr, createItemAcceptor } from "expression";

interface OpenToken {
    type: "open";
    node: ProsemirrorNode;
    createdFromSplit?: boolean;
}

interface CloseToken {
    type: "close";
    node: ProsemirrorNode;
}

interface AcceptedState {
    consumeNode: (node: ProsemirrorNode) => boolean;
    acceptedNodes: ProsemirrorNode[];
}

export type Token = OpenToken | CloseToken;

export const getNaiveTokenList = (node: ProsemirrorNode) => {
    const tokens: Token[] = [];

    const visit = (node: ProsemirrorNode) => {
        tokens.push({ type: "open", node });
        if (node.content) {
            for (const child of node.content) {
                visit(child);
            }
        }
        tokens.push({ type: "close", node });
    };

    visit(node);
    return tokens;
};

export const healNaiveTokenList = (
    tokens: Token[],
    schema: ProsemirrorSchema
): Token[] => {
    const openTokens: OpenToken[] = [];
    const nextTokens: Token[] = [];

    const tokenToAcceptorMap: Map<OpenToken, AcceptedState> = new Map();
    const newOpenTokensMap: Map<ProsemirrorNode, OpenToken[]> = new Map();
    const acceptorExpressions: Map<string, Expr> = new Map();
    const nodeToTokenMap: Map<ProsemirrorNode, OpenToken[]> = new Map();

    Object.entries(schema.nodes).forEach(([key, entry]) => {
        if (entry.spec.content) {
            acceptorExpressions.set(key, parseExpr(entry.spec.content));
        }
    });

    const matchProsemirrorNode = (group: string) => (node: ProsemirrorNode) => {
        const schemaEntry = schema.nodes[node.type];
        const matchRes =
            node.type === group || schemaEntry.spec.group === group;
        return matchRes;
    };

    const openToken = (token: OpenToken) => {
        nextTokens.push(token);
        openTokens.push(token);
        nodeToTokenMap.set(token.node, [
            ...(nodeToTokenMap.get(token.node) || []),
            token,
        ]);
    };

    const getOrAddAcceptedStateToTokenMap = (token): AcceptedState => {
        const existing = tokenToAcceptorMap.get(token);
        if (existing) {
            return existing;
        }
        const next = {
            consumeNode: createItemAcceptor(
                acceptorExpressions.get(token.node.type),
                matchProsemirrorNode
            ),
            acceptedNodes: [] as ProsemirrorNode[],
        };
        tokenToAcceptorMap.set(token, next);
        return next;
    };

    for (const token of tokens) {
        if (token.type === "open") {
            const toReopen: OpenToken[] = [];
            let acceptingParentDepth = openTokens.length - 1;
            let accepted = token === tokens[0];
            while (!accepted) {
                if (acceptingParentDepth < 0) {
                    throw new Error(
                        `Prosemirror healer cannot find a suitable parent node for ${token.node.type}` +
                            ` (closed ${toReopen
                                .concat()
                                .reverse()
                                .map((t) => t.node.type)
                                .join(", ")})`
                    );
                }
                const testingToken = openTokens[acceptingParentDepth];
                const { acceptedNodes, consumeNode } =
                    getOrAddAcceptedStateToTokenMap(testingToken);
                accepted = consumeNode(token.node);
                if (accepted) {
                    acceptedNodes.push(token.node);
                } else {
                    nextTokens.push({
                        type: "close",
                        node: testingToken.node,
                    });
                    openTokens.pop();
                    toReopen.unshift({
                        type: "open",
                        node: testingToken.node,
                    });
                    --acceptingParentDepth;
                }
            }
            openToken(token);
            if (toReopen.length > 0) {
                newOpenTokensMap.set(token.node, toReopen);
            }
        } else if (token.type === "close") {
            nextTokens.push(token);
            openTokens.pop();
            const mustOpenNow = newOpenTokensMap.get(token.node);
            if (mustOpenNow) {
                for (const tokenToOpen of mustOpenNow) {
                    openToken(tokenToOpen);
                }
            }
        }
    }
    return nextTokens.map((token) => {
        const tokensForNode = nodeToTokenMap.get(token.node);
        if (tokensForNode && tokensForNode.length > 1) {
            return { ...token, createdFromSplit: true };
        }
        return token;
    });
};

export const heal = (
    node: ProsemirrorNode,
    schema: ProsemirrorSchema
): ProsemirrorNode => {
    const naiveTokens = getNaiveTokenList(node);
    const tokens = healNaiveTokenList(naiveTokens, schema);
    const parentStack = [];
    let rootNode: ProsemirrorNode;
    for (let i = 0; i < tokens.length; i++) {
        const token = tokens[i];
        const nextToken = tokens[i + 1];
        if (token.type === "open") {
            if (
                nextToken &&
                nextToken.type === "close" &&
                nextToken.node === token.node &&
                token.createdFromSplit
            ) {
                // We found an open(X), close(X) pair that is empty because it was created while
                // spliting an element during healing. Just ignore these nodes.
                ++i;
                continue;
            }
            const contentProp = token.node.content ? { content: [] } : {};
            const nextNode = { ...token.node, ...contentProp };
            if (parentStack.length === 0) {
                rootNode = nextNode;
            } else {
                parentStack[parentStack.length - 1].content.push(nextNode);
            }
            parentStack.push(nextNode);
        } else if (token.type === "close") {
            parentStack.pop();
        }
    }
    if (parentStack.length !== 0) {
        throw new Error(
            "Mismatched tokens encountered while healing document."
        );
    }
    return rootNode;
};
