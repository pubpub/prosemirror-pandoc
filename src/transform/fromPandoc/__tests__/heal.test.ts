/* global describe, it, expect */
import * as prosemirrorSchema from "../../../example/schema";

import { ProsemirrorSchema } from "../../../types";
import { getNaiveTokenList, healNaiveTokenList, Token, heal } from "../heal";

const toyProsemirrorSchmea: ProsemirrorSchema = {
    nodes: {
        A: {
            defining: true,
            content: "B*",
        },
        B: {
            defining: true,
            content: "(C|D)*",
        },
        C: {
            defining: true,
            content: "D*",
        },
        D: {
            defining: true,
            content: "text*",
        },
    },
    marks: {},
};

const stringify = (tokens: Token[]) =>
    tokens
        .map((t: Token) => {
            const prefix = t.type === "open" ? "O" : "C";
            const maybeAsterisk =
                t.type === "open" && t.createdFromSplit ? "*" : "";
            return `${prefix}${maybeAsterisk}(${t.node.type})`;
        })
        .join(" ");

describe("getNaiveTokenList", () => {
    it("tokenizes a Prosemirror tree", () => {
        expect(
            stringify(
                getNaiveTokenList({
                    type: "A",
                    children: [
                        {
                            type: "B",
                            children: [
                                { type: "C", children: [{ type: "D" }] },
                            ],
                        },
                    ],
                })
            )
        ).toEqual("O(A) O(B) O(C) O(D) C(D) C(C) C(B) C(A)");
    });
});

describe("healNaiveTokenList", () => {
    it("passes through a valid tree", () => {
        expect(
            stringify(
                healNaiveTokenList(
                    getNaiveTokenList({
                        type: "A",
                        children: [
                            {
                                type: "B",
                                children: [
                                    { type: "D" },
                                    {
                                        type: "C",
                                        children: [
                                            { type: "D" },
                                            { type: "D" },
                                        ],
                                    },
                                ],
                            },
                        ],
                    }),
                    toyProsemirrorSchmea
                )
            )
        ).toEqual(
            "O(A) O(B) O(D) C(D) O(C) O(D) C(D) O(D) C(D) C(C) C(B) C(A)"
        );
    });

    it("heals an invalid Prosemirror tree", () => {
        const naiveList = getNaiveTokenList({
            type: "A",
            children: [{ type: "B", children: [{ type: "B" }] }],
        });
        expect(
            stringify(healNaiveTokenList(naiveList, toyProsemirrorSchmea))
        ).toEqual("O(A) O*(B) C(B) O(B) C(B) O*(B) C(B) C(A)");
    });

    it("heals a more complicated invalid Prosemirror tree", () => {
        const naiveList = getNaiveTokenList({
            type: "A",
            children: [
                {
                    type: "B",
                    children: [
                        { type: "B" },
                        { type: "D", children: [{ type: "C" }] },
                    ],
                },
            ],
        });
        expect(
            stringify(healNaiveTokenList(naiveList, toyProsemirrorSchmea))
        ).toEqual(
            "O(A) O*(B) C(B) O(B) C(B) O*(B) O*(D) C(D) O(C) C(C) O*(D) C(D) C(B) C(A)"
        );
    });

    it("heals a real-world example of a misplaced image element", () => {
        const naiveList = getNaiveTokenList({
            type: "doc",
            children: [
                {
                    type: "bullet_list",
                    children: [
                        {
                            type: "list_item",
                            children: [
                                {
                                    type: "paragraph",
                                    children: [{ type: "text" }],
                                },
                            ],
                        },
                        { type: "list_item", children: [{ type: "image" }] },
                        {
                            type: "list_item",
                            children: [
                                {
                                    type: "paragraph",
                                    children: [{ type: "text" }],
                                },
                            ],
                        },
                    ],
                },
            ],
        });
        expect(
            stringify(healNaiveTokenList(naiveList, prosemirrorSchema))
        ).toMatchSnapshot();
    });

    it("heals a real-world example of a misplaced image element that is less dire", () => {
        const naiveList = getNaiveTokenList({
            type: "doc",
            children: [
                {
                    type: "bullet_list",
                    children: [
                        {
                            type: "list_item",
                            children: [
                                {
                                    type: "paragraph",
                                    children: [{ type: "text" }],
                                },
                                {
                                    type: "paragraph",
                                    children: [
                                        { type: "text" },
                                        { type: "image" },
                                        { type: "text" },
                                    ],
                                },
                            ],
                        },
                        {
                            type: "list_item",
                            children: [
                                {
                                    type: "paragraph",
                                    children: [{ type: "text" }],
                                },
                            ],
                        },
                    ],
                },
            ],
        });
        expect(
            stringify(healNaiveTokenList(naiveList, prosemirrorSchema))
        ).toMatchSnapshot();
    });
});

describe("heal", () => {
    it("heals an improper Prosemirror document", () => {
        expect(
            heal(
                {
                    type: "doc",
                    children: [
                        {
                            type: "bullet_list",
                            children: [
                                {
                                    type: "list_item",
                                    children: [
                                        {
                                            type: "paragraph",
                                            children: [{ type: "text" }],
                                        },
                                    ],
                                },
                                {
                                    type: "list_item",
                                    children: [{ type: "image" }],
                                },
                                {
                                    type: "list_item",
                                    children: [
                                        {
                                            type: "paragraph",
                                            children: [{ type: "text" }],
                                        },
                                    ],
                                },
                            ],
                        },
                    ],
                },
                prosemirrorSchema
            )
        ).toEqual({
            type: "doc",
            children: [
                {
                    type: "bullet_list",
                    children: [
                        {
                            type: "list_item",
                            children: [
                                {
                                    type: "paragraph",
                                    children: [{ type: "text" }],
                                },
                            ],
                        },
                    ],
                },
                { type: "image" },
                {
                    type: "bullet_list",
                    children: [
                        {
                            type: "list_item",
                            children: [
                                {
                                    type: "paragraph",
                                    children: [{ type: "text" }],
                                },
                            ],
                        },
                    ],
                },
            ],
        });
    });

    it("heals a broken Prosemirror document that requires less intervention", () => {
        expect(
            heal(
                {
                    type: "doc",
                    children: [
                        {
                            type: "bullet_list",
                            children: [
                                {
                                    type: "list_item",
                                    children: [
                                        {
                                            type: "paragraph",
                                            children: [{ type: "text" }],
                                        },
                                        {
                                            type: "paragraph",
                                            children: [
                                                { type: "text" },
                                                { type: "image" },
                                                { type: "text" },
                                            ],
                                        },
                                    ],
                                },
                                {
                                    type: "list_item",
                                    children: [
                                        {
                                            type: "paragraph",
                                            children: [{ type: "text" }],
                                        },
                                    ],
                                },
                            ],
                        },
                    ],
                },
                prosemirrorSchema
            )
        ).toEqual({
            type: "doc",
            children: [
                {
                    type: "bullet_list",
                    children: [
                        {
                            type: "list_item",
                            children: [
                                {
                                    type: "paragraph",
                                    children: [{ type: "text" }],
                                },
                                {
                                    type: "paragraph",
                                    children: [{ type: "text" }],
                                },
                                { type: "image" },
                                {
                                    type: "paragraph",
                                    children: [{ type: "text" }],
                                },
                            ],
                        },
                        {
                            type: "list_item",
                            children: [
                                {
                                    type: "paragraph",
                                    children: [{ type: "text" }],
                                },
                            ],
                        },
                    ],
                },
            ],
        });
    });
});
