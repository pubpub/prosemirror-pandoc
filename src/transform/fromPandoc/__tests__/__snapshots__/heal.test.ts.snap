// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`healNaiveTokenList heals a real-world example of a misplaced image element 1`] = `"O(doc) O*(bullet_list) O(list_item) O(paragraph) O(text) C(text) C(paragraph) C(list_item) O*(list_item) C(list_item) C(bullet_list) O(image) C(image) O*(bullet_list) O*(list_item) C(list_item) O(list_item) O(paragraph) O(text) C(text) C(paragraph) C(list_item) C(bullet_list) C(doc)"`;

exports[`healNaiveTokenList heals a real-world example of a misplaced image element that is less dire 1`] = `"O(doc) O(bullet_list) O(list_item) O(paragraph) O(text) C(text) C(paragraph) O*(paragraph) O(text) C(text) C(paragraph) O(image) C(image) O*(paragraph) O(text) C(text) C(paragraph) C(list_item) O(list_item) O(paragraph) O(text) C(text) C(paragraph) C(list_item) C(bullet_list) C(doc)"`;
